compiler/damlc/tests/src/DA/Test/ShakeIdeClient.hs

```diff
@@ -510,6 +498,45 @@ dlintSmokeTests mbScenarioService = Tasty.testGroup "Dlint smoke tests"
             setFilesOfInterest [foo]
             expectNoErrors
             expectDiagnostic DsInfo (foo, 2, 8) "Suggestion: Redundant case"
+    ,  testCase' "Use let" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo g x = do"
+                , "  y <- pure x"
+                , "  g y"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 10) "Suggestion: Use let"
+    -- `forA_` and friends not implemented yet : https://github.com/digital-asset/daml/issues/3315
+    -- ,  testCase' "Redundant void" $ do
+    --         foo <- makeFile "Foo.daml" $ T.unlines
+    --             [ "daml 1.2"
+    --             , "module Foo where"
+    --             , "foo g xs = void $ forA_ g xs"
+    --             ]
+    --         setFilesOfInterest [foo]
+    --         expectNoErrors
+    --         expectDiagnostic DsInfo (foo, XX, XX) "Suggestion: Redundant void"
+    ,  testCase' "Use <$>" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo f g bar = do x <- bar; return (f $ g x)"
```

Comment:
I wonder if this rule and the one after are a good idea for DAML. Currently we are mostly encouraging people to use `do` notation and not worry about this Functor, Applicative, Monad thing and these rules seem to go in the other direction.

What do you think @hurryabit @bame-da